cloudinit/util.py

```diff
@@ -68,17 +68,31 @@
                    ['running-in-container'],
                    ['lxc-is-container'])
 
-PROC_CMDLINE = None
+_CACHED_RESPONSES = {}
 
-_LSB_RELEASE = {}
 
+def memoize(f):
+    """
+    Function used to cache responses from functions decorated with it.
+    """
+
+    def helper(*x):
+        global _CACHED_RESPONSES
+        if (f, x) not in _CACHED_RESPONSES:
```

Comment:
This will cause errors at runtime if the arguments to the function aren't hashable (the most obvious case being if a dict is passed):

```py
>>> def foo(): pass                                                                                                                                                                                            
>>> (foo, ({},)) in {}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'dict'
```

At the very least we should capture this in the docstring so that future developers understand the caveat.

However, given that only one of the functions being decorated takes an argument, I wonder if we might be better off simplifying this so that it only memoize's the zero-argument response and always passes through if an argument is specified. To express that in (completely untested :3) code, something like:

```py
def helper(*x):
    if x:
        return f(*x)
    if f not in _CACHED_RESPONSES:
        _CACHED_RESPONSES[f] = f()
    return _CACHED_RESPONSES[f]
```
```